File: AutoChess/assets/script/config/skill_config.ts
/*
 * skill_config.ts
 * author: qianqians
 * 2023/10/2
 */
import { JsonAsset, resources, error } from 'cc';
import { ChangePositionType, SwapPropertiesType } from '../other/enums';
import { Direction, Priority } from '../serverSDK/common';

export class SkillConfig {
    public Id: number;
    public Name: string;
    public Type: string;
    public Priority: Priority;
    public EffectTime: number;
    public Effect: number;
    public ObjCount: number;
    public ObjectDirection: Direction;
    public Level1Value_1: number;
    public Level1Value_2: number;
    public Level2Value_1: number;
    public Level2Value_2: number;
    public Level3Value_1: number;
    public Level3Value_2: number;
    public SummonId: number[];
    public SummonLevel: number;
    public ChangePositionType : ChangePositionType;
    public SwapPropertiesType : SwapPropertiesType;
    public AddBufferID: number;
}

export async function LoadSkillConfig() : Promise<Map<number, SkillConfig>> {
    return new Promise<Map<number, SkillConfig>>((resolve, reject)=>{
        let map = new Map<number, SkillConfig>();

        console.log("Load Skill Config begin!");
        resources.load('config/Skill', (err: any, res: JsonAsset) => {
            if (err) {
                error(err.message || err);
                return;
            }
            // 获取到 Json 数据
            const jsonData: object = res.json!;
            Object.keys(jsonData).forEach((k) => {
                let v = jsonData[k];

                let skillc = new SkillConfig();
                skillc.Id = v["Id"];
                skillc.Name = v["Name"];
                skillc.Type = v["Type"];
                skillc.Priority = Priority[v["Priority"] as keyof typeof Priority];
                skillc.EffectTime = v["EffectTime"];
                skillc.Effect = v["Effect"];
                skillc.ObjCount = v["ObjCount"];
                skillc.ObjectDirection = v["ObjDirection"];
                skillc.Level1Value_1 = v["Level1Value_1"];
                skillc.Level1Value_2 = v["Level1Value_2"];
                skillc.Level2Value_1 = v["Level2Value_1"];
                skillc.Level2Value_2 = v["Level2Value_2"];
                skillc.Level3Value_1 = v["Level3Value_1"];
                skillc.Level3Value_2 = v["Level3Value_2"];
                skillc.SummonId = v["SummonId"] == "" ? [] : JSON.parse(v["SummonId"]);
                skillc.SummonLevel = v["SummonLevel"];
                skillc.ChangePositionType = v["ChangePositionType"];
                skillc.SwapPropertiesType = v["SwapPropertiesType"];
                skillc.AddBufferID =  v["AddBufferID"];

                map.set(parseInt(k), skillc);
            });

            console.log("Load Skill Config end!");
            resolve(map);
        });
    });
}

File: server/server/config/skill_config.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Reflection.Metadata;

namespace config
{
    public class SkillConfig
    {
        public int Id;
        public string Name;
        public string Type;
        public string Priority;
        public int EffectTime;
        public int Effect;
        public int ObjCount;
        public int ObjectDirection;
        public int Level1Value_1;
        public int Level1Value_2;
        public int Level2Value_1;
        public int Level2Value_2;
        public int Level3Value_1;
        public int Level3Value_2;
        public string SummonId;
        public int SummonLevel;
        public int ChangePositionType;
        public int SwapPropertiesType;
        public int AddBufferID;

        public static Dictionary<int, SkillConfig> Load(string path)
        {
            var obj = new Dictionary<int, SkillConfig>();

            FileStream fs = File.OpenRead(path);
            byte[] data = new byte[fs.Length];
            int offset = 0;
            int remaining = data.Length;
            while (remaining > 0)
            {

                int read = fs.Read(data, offset, remaining);
                if (read <= 0)
                    throw new EndOfStreamException("file read at" + read.ToString() + " failed");

                remaining -= read;
                offset += read;
            }

            var handle = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, dynamic>>(System.Text.Encoding.Default.GetString(data));
            foreach (var o in handle.Values)
            {
                var skillc = new SkillConfig();
                skillc.Id = (int)o["Id"];
                skillc.Name = (string)o["Name"];
                skillc.Type = (string)o["Type"];
                skillc.Priority = (string)o["Priority"];
                skillc.EffectTime = (int)o["EffectTime"];
                skillc.Effect = (int)o["Effect"];
                if (o["ObjCount"] != null)
                {
                    skillc.ObjCount = (int)o["ObjCount"];
                }
                if (o["ObjectDirection"] != null)
                {
                    skillc.ObjectDirection = (int)o["ObjectDirection"];
                }
                if (o["Level1Value_1"] != null)
                {
                    skillc.Level1Value_1 = (int)o["Level1Value_1"];
                }
                if (o["Level1Value_2"] != null)
                {
                    skillc.Level1Value_2 = (int)o["Level1Value_2"];
                }
                if (o["Level2Value_1"] != null)
                {
                    skillc.Level2Value_1 = (int)o["Level2Value_1"];
                }
                if (o["Level2Value_2"] != null)
                {
                    skillc.Level2Value_2 = (int)o["Level2Value_2"];
                }
                if (o["Level3Value_1"] != null)
                {
                    skillc.Level3Value_1 = (int)o["Level3Value_1"];
                }
                if (o["Level3Value_2"] != null)
                {
                    skillc.Level3Value_2 = (int)o["Level3Value_2"];
                }
                if (o["SummonId"] != null)
                {
                    skillc.SummonId = (string)o["SummonId"];
                }
                if (o["SummonLevel"] != null)
                {
                    skillc.SummonLevel = (int)o["SummonLevel"];
                }
                if (o["ChangePositionType"] != null)
                {
                    skillc.ChangePositionType = (int)o["ChangePositionType"];
                }
                if (o["SwapPropertiesType"] != null)
                {
                    skillc.SwapPropertiesType = (int)o["SwapPropertiesType"];
                }
                if (o["AddBufferID"] != null)
                {
                    skillc.AddBufferID = (int)o["AddBufferID"];
                }
                obj[skillc.Id] = skillc;
            }

            return obj;
        }
    }
}